network/retrieval/peer.go

```diff
@@ -17,20 +17,67 @@
 package retrieval
 
 import (
+	"bytes"
+	"encoding/hex"
+	"errors"
+	"sync"
+
 	"github.com/ethereum/go-ethereum/log"
 	"github.com/ethersphere/swarm/network"
+	"github.com/ethersphere/swarm/storage"
 )
 
-// Peer wraps BzzPeer with a contextual logger for this peer
+// retrieval represents an open, ongoing retrieval sent to
+// a specific peer
+type retrieval struct {
+	ruid uint
+	addr storage.Address
+}
+
+// Peer wraps BzzPeer with a contextual logger and tracks open
+// retrievals for that peer
 type Peer struct {
-	*network.BzzPeer
-	logger log.Logger
+	mtx              sync.Mutex         // synchronize retrievals
+	*network.BzzPeer                    // bzzpeer
+	logger           log.Logger         // logger with base and peer address
+	retrievals       map[uint]retrieval // current ongoing retrievals
 }
 
 // NewPeer is the constructor for Peer
-func NewPeer(peer *network.BzzPeer) *Peer {
+func NewPeer(peer *network.BzzPeer, baseKey []byte) *Peer {
 	return &Peer{
-		BzzPeer: peer,
-		logger:  log.New("peer", peer.ID()),
+		BzzPeer:    peer,
+		logger:     log.New("base", hex.EncodeToString(baseKey)[:16], "peer", peer.ID().String()[:16]),
+		retrievals: make(map[uint]retrieval),
+	}
+}
+
+// chunkRequested adds a new retrieval to the retrievals map
+// this is in order to identify unsolicited chunk deliveries
+func (p *Peer) chunkRequested(ruid uint, addr storage.Address) {
+	p.mtx.Lock()
+	defer p.mtx.Unlock()
+	ret := retrieval{
+		ruid: ruid,
+		addr: addr,
 	}
+	p.retrievals[ruid] = ret
+}
+
+// chunkReceived is called upon ChunkDelivery message reception
+// it is meant to idenfify unsolicited chunk deliveries
+func (p *Peer) chunkReceived(ruid uint, addr storage.Address) error {
+	p.mtx.Lock()
+	defer p.mtx.Unlock()
+	v, ok := p.retrievals[ruid]
+	if !ok {
+		return errors.New("unsolicited chunk delivery - cannot find ruid")
+	}
+	if !bytes.Equal(v.addr, addr) {
+		return errors.New("retrieve request found but address does not match")
```

Comment:
i don't think there's a potential leak here since the error will cause the peer to get dropped so it actually doesn't really matter that much. in any case i added a `defer delete` statement above instead of just deleting when we receive the _correct_ request

Reply:
Yes, you are right. Thanks fo the change.